cumulus/polkadot-parachain/src/service.rs

```diff
@@ -1410,7 +1410,9 @@ where
 	<<AuraId as AppCrypto>::Pair as Pair>::Signature:
 		TryFrom<Vec<u8>> + std::hash::Hash + sp_runtime::traits::Member + Codec,
 {
-	start_node_impl::<RuntimeApi, _, _, _>(
+	// TODO: is this right? Seems okay since asset hub node is a "shell" node. Can Kusama start
+	// using this right away?
+	start_basic_lookahead_node_impl::<RuntimeApi, _, _, _>(
```

Comment:
I wasn't sure about this because asset-hub-kusama also calls this line. Would it break something? I would guess we should we test this change on rococo first actually.

Reply:
I don't know this node code well enough, but my understanding is that it needs the `lookahead`?

Reply:
You're right, Kusama and Polkadot AH both call this too. A change here will affect how `polkadot-parachain` runs any of them, so if you want to only move testnets over in the next released version without moving Kusama and Polkadot AH onto async backing you'll need two separate functions.

Reply:
Thanks! So I'll copy/paste this and `start_node_impl` into new functions e.g. `start_lookahead_node_impl` with the necessary changes from the guide. It's a bit ugly but I guess soon the non-async-backing `start_asset_hub_node` can be removed.

Reply:
@seadanda Looks like `AuraParams` is missing `collation_request_receiver`, which is `Some` in this case. Should I add it, and copy the code that uses it from `basic_aura` to `lookahead::aura`?

Reply:
Yea it would be nice to find a cleaner way to combine them, the diff is very small between the four `_impl` functions, but nothing jumps out at me immediately because of the slightly different constraints.

Not too sure on your second comment, I'm not familiar with the implementation of lookahead vs basic.

Reply:
I'm getting a really long error message now. 

```
error[E0277]: the trait bound `<RuntimeApi as ConstructRuntimeApi<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>, sc_service::client::client::Client<sc_client_db::Backend<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, LocalCallExecutor<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>, sc_client_db::Backend<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, WasmExecutor<(sp_io::storage::HostFunctions, sp_io::default_child_storage::HostFunctions, sp_io::misc::HostFunctions, sp_io::wasm_tracing::HostFunctions, sp_io::offchain::HostFunctions, sp_io::crypto::HostFunctions, sp_io::hashing::HostFunctions, sp_io::allocator::HostFunctions, sp_io::panic_handler::HostFunctions, sp_io::logging::HostFunctions, sp_io::trie::HostFunctions, sp_io::offchain_index::HostFunctions, sp_io::transaction_index::HostFunctions)>>, sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>, RuntimeApi>>>::RuntimeApi: cumulus_primitives_aura::AuraUnincludedSegmentApi<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>` is not satisfied
    --> cumulus/polkadot-parachain/src/service.rs:1864:80
     |
1864 |                 aura::run::<Block, <AuraId as AppCrypto>::Pair, _, _, _, _, _, _, _, _, _>(params)
     |                 -------------------------------------------------------------------------- ^^^^^^ the trait `cumulus_primitives_aura::AuraUnincludedSegmentApi<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>` is not implemented for `<RuntimeApi as ConstructRuntimeApi<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>, sc_service::client::client::Client<sc_client_db::Backend<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, LocalCallExecutor<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>, sc_client_db::Backend<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, WasmExecutor<(sp_io::storage::HostFunctions, sp_io::default_child_storage::HostFunctions, sp_io::misc::HostFunctions, sp_io::wasm_tracing::HostFunctions, sp_io::offchain::HostFunctions, sp_io::crypto::HostFunctions, sp_io::hashing::HostFunctions, sp_io::allocator::HostFunctions, sp_io::panic_handler::HostFunctions, sp_io::logging::HostFunctions, sp_io::trie::HostFunctions, sp_io::offchain_index::HostFunctions, sp_io::transaction_index::HostFunctions)>>, sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>, RuntimeApi>>>::RuntimeApi`
     |                 |
     |                 required by a bound introduced by this call
     |
note: required by a bound in `cumulus_client_consensus_aura::collators::lookahead::run`
    --> /Users/marcinparity/Repos/github.com/paritytech/polkadot-sdk-3/cumulus/client/consensus/aura/src/collators/lookahead.rs:127:61
     |
113  | pub fn run<Block, P, BI, CIDP, Client, Backend, RClient, CHP, SO, Proposer, CS>(
     |        --- required by a bound in this function
...
127  |         AuraApi<Block, P::Public> + CollectCollationInfo<Block> + AuraUnincludedSegmentApi<Block>,
     |                                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `run`
help: consider further restricting the associated type
     |
1751 |         TryFrom<Vec<u8>> + std::hash::Hash + sp_runtime::traits::Member + Codec, <RuntimeApi as ConstructRuntimeApi<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>, sc_service::client::client::Client<sc_client_db::Backend<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, LocalCallExecutor<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>, sc_client_db::Backend<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, WasmExecutor<(sp_io::storage::HostFunctions, sp_io::default_child_storage::HostFunctions, sp_io::misc::HostFunctions, sp_io::wasm_tracing::HostFunctions, sp_io::offchain::HostFunctions, sp_io::crypto::HostFunctions, sp_io::hashing::HostFunctions, sp_io::allocator::HostFunctions, sp_io::panic_handler::HostFunctions, sp_io::logging::HostFunctions, sp_io::trie::HostFunctions, sp_io::offchain_index::HostFunctions, sp_io::transaction_index::HostFunctions)>>, sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>, RuntimeApi>>>::RuntimeApi: cumulus_primitives_aura::AuraUnincludedSegmentApi<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>
     |                                                                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I think in essence it's saying that `AuraUnincludedSegmentApi<Block>` bound is missing for `RuntimeApi`, but I'm lost on how to fix it. Maybe after the transition to aura, a new client needs to be constructed?

Reply:
@georgepisaltu Do you have any ideas on this or https://github.com/paritytech/polkadot-sdk/pull/2826/#pullrequestreview-1816111722?

Reply:
For this long error, you just need to add the correct trait bound, `cumulus_primitives_aura::AuraUnincludedSegmentApi<Block>`, [here](https://github.com/paritytech/polkadot-sdk/blob/f7306d327e0eee0a641ac43628ab525b0096ba7c/cumulus/polkadot-parachain/src/service.rs#L1409).

In the case of AH collators, the runtime initially started as shell and then later transitioned into the real runtime which had Aura. Therefore, because the genesis is already set and will never change, AH collators will always have to account for this edge case, meaning they `start_asset_hub_node` will always be different from the rest. More details [here](https://github.com/paritytech/polkadot-sdk/issues/1566) and [here](https://github.com/paritytech/polkadot-sdk/pull/1788).

The `collation_request_receiver` was added because there is already a relay chain collation request stream initialized for the shell consensus, and it can't be initialized again so it needs to be reused when transitioning into Aura. More details [here](https://github.com/paritytech/polkadot-sdk/pull/1911).